src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -413,8 +413,38 @@ public List<string> PullFiles(string devicePath, string localPath)
         }
 
 
+
         public string? GetDeviceToUse(ILogger logger, string apkRequiredProperty, string propertyName)
         {
+            var allDevicesAndTheirProperties = GetAllDevicesToUse(logger, apkRequiredProperty, propertyName);
+            if (allDevicesAndTheirProperties.Count > 0)
+            {
+                var firstAvailableCompatible = allDevicesAndTheirProperties.First();
+                logger.LogDebug($"Using first-found compatible device of {allDevicesAndTheirProperties.Count} total- serial: '{firstAvailableCompatible.Key}' - {propertyName}: {firstAvailableCompatible.Value}");
+                return firstAvailableCompatible.Key;
+            }
+            return null;
+        }
+
+        public string? GetUniqueDeviceToUse(ILogger logger, string apkRequiredProperty, string propertyName)
+        {
+            var devices = GetAllDevicesToUse(logger, apkRequiredProperty, propertyName);
+            if (devices.Count == 0)
+            {
+                logger.LogError("Cannot find a device with app installed, please check that a device is attached or run install first");

```

Comment:
I think this error is too custom for the purpose of this method. I think the calling method knows it's looking for the app installed but this one only looks for any property?

Reply:
Yeah, looking lower at the usage - it's the caller of this who know we're looking for a "device with a specific app". I would consider removing this method and just doing the logic there (find all devices and if there's >1, throw the error from there). If it's used in multiple places, you could have an `AndroidCommandBase` that has this logic. But this is just my opinion and I will leave it on your decision. Fixing the error message is also fine

Reply:
I choose fixing the message 😃